specs/carbon-ql-k8s-specification.md

```diff
@@ -0,0 +1,357 @@
+```
+
+
+## Kubernetes components
+
+The Cabon-QL API includes the following components:
+
+Admin scenario:
+- Node (smallest infra component, aka foundational component)
+- Node Pool
+- Cluster
+
+Workload scenario:
+- Pod (smalled workload component, aka foundation component)
+- Deployment
+- Job
+
+## Kuberetnes components composition
+
+Carbon-QL components composition patterns for K8s:
+
+- **single foundational components (pod, node, disk,..)** : defines the base model to use, the params & telemetry, can set default vars
+
+```yaml 
+
+vars : <default vars>
+model:
+  path: azure.computte.aks.node.sci
+  params : <model params>
+  telemetry : <model telemetry>
+```
+ ---
+
+
+- **calculating SCI from subscomponents** : use sub components to calculate their SCI core, and defines the model to compute the resulting SCI (addition subcomponents SCI in most cases)
+
+```yaml 
+model:
+  path: gsf.common.AdditionSubComponentsSCIs:v1
+  params : ~
+  telemetry : ~
+  subComponents:
+    <SubComponentKind>
+        - name: <another component same as above>
+          metadata : 
+          params:
+          path: 
+          telemetry: 
+```
+
+the current components, when calling subcomponents, can override 
+          - metadata : 
+          - params:
+          - path: 
+          - telemetry: 
+
+
+
+----
+
+
+- **calculating SCI based on a attribution % from another SCI** : to calculate SCI for workloads and shared environments, an SCI for a pod, would be attributed  (or calculated) from the SCI of the hosting node (based on % of node resources consumed by the pod for the defined time duration)
+
+```yaml
+
+model:
+  path: azure.computte.aks.pod.sci
+  params:
+    <key>: <value>
+telemetry:
+  - <telemetry data>
+SubComponents: ~
+SCIAttributedFrom:
+  - SCIComponentKind: k8sNode (or point to SCIComponentName directly mynode1)
+    SCIComponentparams: 
+      nodeName : <node name>
+
+```
+
+
+
+## Scenarios
+
+## Example for Admin scenarios 
+
+An admin wants to calculate the infrastructure SCI for an AKS cluster in Azure. They can use the following components:
+
+- Node: my-node
+- Node Pool: my-node-pool
+- Cluster: my-cluster
+
+
+```yaml
+kind: k8sCluster
+name : AKS-PROD-1
+metadata: 
+  attributes: compute
+vars:
+  vendor: azure
+  region: eastus
+  sku: standard
+  rg : rg-prod-1
+model:
+  path: azure.aks.cluster.sci:v1
+  params:
+    clusterSKU: @vars.sku
+    clusterName : @name
+    clusterResourceGroup : @vars.rg
+  telemetry : ~
+  SubComponents:
+      k8sNodePool:
+        - name: mynode_pool-1
+          metadata:
+              app: batch
+        - name: mynode_pool-2
+          metadata:
+              app: web
+
+```
+
+### Example for Developer scenario
+A developer wants to calculate the workload SCI for a deployment and a job in an AKS cluster in Azure. 
+
+They can use the following components:
+
+- Pod: my-pod
+- Deployment: my-deployment
+- Job: my-job
+- Namespace: my-namespace
+
+
+```yaml
```

Comment:
Again thinking about how this maps to the current paradigm (summation vs. attributional)

We atomically calculate the emissions of each component using a model and some observations and these are summed up to get the emissions for the entire workload.

```yaml
name: my-azure-workload
tags:
  attributes: compute
  vendor: azure
  region: eastus
  AKSClusterRG: rg-prod-1
  AKSClusterNanme: AKS-PROD-1
  workloadNamespace: AKS-PROD-1
components:
    - name: k8sDeployment # is this the component for administering the workload?
      tags:
          attributes: compute
      model: 
        path: azure.aks.node.sci:v1
      config:
        deploymentNamespace : @vars.workloadNamespace
        clusterName: @vars.AKSClusterName
        cluster
        AKSClusterRG: @vars.AKSClusterRG
      observations:
        ???                        
    - name: my-job
      model: 
        path: azure.aks.job.sci:v1    
      observations:
        ???
    - name: my-cache
      model: 
        path: azure.compute.redis.sci:v1    
      observations:
        ???
```

Reply:
thanks a lot @jawache for taking the time to look closely at this proposal. 

I think we're aligned on the summation paradigm. 

Now the attributional paradigm (or the allocation paradigm) can be generalized to any shared hosting scenario (a node shared by pods on Kubernetes, a DB server shared by multiple DBs, a VM shared by multiple web servers, etc.)

Aligned that we need to build up SCI from the smallest component to the englobing component (my workload). 

now pods on Kub are the smallest workload component (atomic) ; i think we could say the same for a DB instance (so my workload SCI would use SCI summation paradigm of fhe Front + API + DB  SCIs)

I think we have 2 options to address this scenario  : (just brainstorming)

1 - c.f your comment above, we just feed the data to the pod model --> however in that case the pod model would be "responsible" for handling pod telemetry (itself) + node telemetry (external to the model)

"""A pod model needs to have as input some observations about it's running environment (the node) and the model needs to be able to return the emissions of just this pod running on all the nodes it's running on.
So I would switch this around to asking what observations does you need to make regarding a pod in order to calculate it's emissions atomically.""""

--
2 - we explicitly add a paradigm to the spec allowing to caculate the SCI of an (atomical) component as a allocation (% or using some other model) of the SCI of another component (for shared hosting envs)

It reminds me of the cost allocation rules (for shared envs on Azure) : https://learn.microsoft.com/en-us/azure/cost-management-billing/costs/allocate-costs

as a simple scenario, we could imagine an SCI allocation rule (maybe better name than attribution) that is STATIC (using a static % allocation ; we aim for better models, but static allocation rules might be handy when no real telemetry is not available)

```yaml
name: all-pods-of-my-API-service
components:
    - name: pod-type-1
      model: 
        path: azure.aks.pod.sci:v1  --> is this scenarios: this pod model uses pod obersvations + allocation rules from the node SCI, to caculate the SCI of the pod
      observations:
        - 2023-07-06T00:00:
          duration: 5
          cpu: 0.34
        - 2023-07-06T00:05:
          duration: 5
          cpu: 0.23
     allocations
        - 2023-07-06T00:05:
          sourceComponent: Mynode1Component (deployed on availability zone 1 / Data center 1)
          model: %allocation
             params:
                  %allocation = 20        
    - name: pod-type-2
      model: 
        path: azure.aks.pod.sci:v1
      observations:
        - 2023-07-06T00:00:
          duration: 5
          cpu: 0.34
        - 2023-07-06T00:05:
          duration: 5
          cpu: 0.23
     allocations
        - 2023-07-06T00:05:
          sourceComponent: Mynode2Component (on availability zone 2)
          model: %allocation
             params:
                  %allocation = 20        

```
and once we have the SCI of the atomical component, we use summation paradigm to build up to our workload SCI.

Ps: i think also the spec should define a kind or type, this would allow to use component templates that define the required params, data types, impact type (SCI, GHG, water, etc.)

    - name: pod-type-1
    - name: pod-type-2
    
    -> kind : pod
        name : pod1
        
 pss: we could go further (later in the roadmap) to allow component "inheritence":
 
 we have a base component VM
 with chid components Azure VM, AWS, GCP, Onpremise that would use different params or different names for the same param